File: lr/src/questions/q_0261_graph_valid_tree.rs
pub fn valid_tree(n: i32, edges: Vec<Vec<i32>>) -> bool {
    let mut graph = vec![vec![]; n as usize];
    for edge in edges {
        graph[edge[0] as usize].push(edge[1]);
        graph[edge[1] as usize].push(edge[0]);
    }
    let mut visited = vec![false; n as usize];
    if !dfs(&graph, &mut visited, 0, -1) {
        return false;
    }
    for i in 0..n {
        if !visited[i as usize] {
            return false;
        }
    }
    true
}

fn dfs(graph: &Vec<Vec<i32>>, visited: &mut Vec<bool>, node: i32, parent: i32) -> bool {
    if visited[node as usize] {
        return false;
    }
    visited[node as usize] = true;
    for &neighbor in graph[node as usize].iter() {
        if neighbor != parent && !dfs(graph, visited, neighbor, node) {
            return false;
        }
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_261() {
        assert_eq!(
            valid_tree(5, vec![vec![0, 1], vec![0, 2], vec![0, 3], vec![1, 4]]),
            true
        );
        assert_eq!(
            valid_tree(
                5,
                vec![vec![0, 1], vec![1, 2], vec![2, 3], vec![1, 3], vec![1, 4]]
            ),
            false
        );
    }
}

File: lg/src/questions/q_0261_graph_valid_tree.go
package questions

func ValidTree(n int, edges [][]int) bool {
	neighbor := make(map[int][]int)
	visited := make(map[int]struct{})
	for _, edge := range edges {
		neighbor[edge[0]] = append(neighbor[edge[0]], edge[1])
		neighbor[edge[1]] = append(neighbor[edge[1]], edge[0])
	}

	var dfs func(node int, parent int) bool
	dfs = func(node int, parent int) bool {
		if _, ok := visited[node]; ok {
			return false
		}
		visited[node] = struct{}{}
		for _, nei := range neighbor[node] {
			if nei != parent && !dfs(nei, node) {
				return false
			}
		}
		return true
	}
	if !dfs(0, -1) {
		return false
	}

	for i := 0; i < n; i++ {
		if _, ok := visited[i]; !ok {
			return false
		}
	}

	return true
}
